### DISCUSSION TITLE
Front end and extension licensing clarifications

### DISCUSSION THREAD

**Discussion by MikkoSteerpath** [directus/directus#17218] [Feature Requests]
IANAL, but:

Currently it appears the whole Directus code base, including front end and code executing extensions, is under a GPL 3.0 license. 

For extensions:

The current implicit understanding from blog posts by directus is that extensions could be under different licenses, however it looks like the extensions SDK is licensed under GPL 3.0 (https://github.com/directus/directus/blob/main/packages/extensions-sdk/) thus making any extension also become licensed under GPL 3.0

For front end code:

It appears the entire front end code is also licensed under GPL 3.0. When front end code is bundled (compiled) and served by the backend to be executed in the end-users browser (distrubted), it can be argued that any and all portions of the front end are subject to the GPL 3.0 license and those rights are given to any end user. This would make it impossible to do any customization of the front end without giving end users the rights to the source code of said customizations. 

Perhaps the license be changed to for example be "GPL 3.0 + exceptions", similarly how the LGPL 3.0 license is constructed?

**Comment by rijkvanzanten** [comment]
IANAL too, but as far as I understand it at the moment there's a difference between usage, modification, and redistribution. I agree that the extensions-sdk should probably be licensed more permissively to ensure there's no gray areas there, but when it comes to the app though, yes you are correct: customizations of the source codebase require the adjustments to be open source and GPL as well. However, app extensions are not modifying the app code itself in any way, so "modifying" the app by installing extensions isn't modifying the codebase as well, as the app bundle itself isn't modified at all when extensions are used 🤔

**Comment by MikkoSteerpath** [comment]
So are you saying that for the front end components written in Vue, all custom components (say we replace the map provider with a different one) would need to be GPL 3.0 as well? That would make it not possible to use for exmaple the Google Maps SDK for the maps, as that is under a non-GPL 3.0 compatible license. 

As I understand it, GPL is an end-user license, for the entirety of the compiled (front end) software including dependencies. When distributed as a whole and intended to be used as a whole, and executed on the end-user system (like the front end code is), the whole thing needs to be GPL 3.0. Meaning the end user has the right to ask for the full source code needed, including the source code of Google Maps SDK.

**Reply by rijkvanzanten** [reply]
To me, and very much IANAL, there's a difference between using the prod build + an extension and making a fork of the source code, and making alterations to it. In the latter, yes it's required to be open source GPL. For the former, the extension itself should be able to be licensed however you want.

**Comment by MikkoSteerpath** [comment]
As far as how I have understood the GPL license, unless there is an exception to the license, the whole distributed software is licensed to the end user under the GPL, and thus not allowing the extension to remain closed source. As far as I understood this discussion, it sounds like you would wish to have the case GPL + exception to allow for the extension to be licensed under another license. GPL does by default not have such a provision.

**Reply by rijkvanzanten** [reply]
> As far as how I have understood the GPL license, unless there is an exception to the license, the whole distributed software is licensed to the end user under the GPL

That's the part I'm not sure about 🙂 Extensions are currently distributed separately from the "base" software and dynamically linked in runtime, so I'm wondering how that changes the "distributed" nature here 🤷🏻 (in any case, yes, we probably should have a very explicit exception for extensions to make sure this confusion is taken care of once and for all)

**Reply by badrange** [reply]
Hi @rijkvanzanten I suggest taking a look at Drupal's licensing FAQ; the situation is quite similar except for the fact that Drupal is licensed under "GPL 2 or later". I think (would be willing to bet on it, but haven't checked) they have gotten help from legal experts when writing that page.

https://www.drupal.org/about/licensing

**Reply by lexoyo** [reply]
I always thought a plugin was using the GPL-lisenced API so had to be GPL but the link you provided @badrange says

> the GPL does not restrict the use of code under incompatible licenses, only the packaging or distribution of it with GPL software. Drupal.org cannot host this incompatible code, but installing those dependencies with a tool like Composer is okay.

**Reply by badrange** [reply]
> I always thought a plugin was using the GPL-lisenced API so had to be GPL but the link you provided @badrange says
> 
> > the GPL does not restrict the use of code under incompatible licenses, only the packaging or distribution of it with GPL software. Drupal.org cannot host this incompatible code, but installing those dependencies with a tool like Composer is okay.

I think you might be thinking about the AGPL license @lexoyo

https://snyk.io/learn/agpl-license/

**Comment by DracoBlue** [comment]
IANAL, just quoting the drupal link:

> > If I write a module or theme, do I have to license it under the GPL?
>
> Yes. Drupal modules and themes are a derivative work of Drupal. If you distribute them, you must do so under the terms of the GPL version 2 or later. You are not required to distribute them at all, however. (See question 8 below.)
> 
> However, when distributing your own Drupal-based work, it is important to keep in mind what the GPL applies to. The GPL on code applies to code that interacts with that code, but not to data. That is, Drupal's PHP code is under the GPL, and so all PHP code that interacts with it must also be under the GPL or GPL compatible. Images, JavaScript, and Flash files that PHP sends to the browser are not affected by the GPL because they are data. However, Drupal's JavaScript, including the copy of jQuery that is included with Drupal, is itself under the GPL as well, so any Javascript that interacts with Drupal's JavaScript in the browser must also be under the GPL or a GPL compatible license.
> 
> When distributing your own module or theme, therefore, the GPL applies to any pieces that directly interact with parts of Drupal that are under the GPL. Images and Flash files you create yourself are not affected. However, if you make a new image based off of an image that is provided by Drupal under the GPL, then that image must also be under the GPL.
> 
> If you commit that module or theme to a Drupal Git repository, however, then all parts of it must be under the GPL version 2 or later, and you must provide the source code. That means the editable form of all files, as described above.

GPL is afaik about distributing something which has GPL in it. If you don’t want to have the GPL effect: use LGPL (like libc does) to make distribution possible. If distribution shall also be GPLing the stuff - use AGPL. 

IMHO as not being a lawyer - the frontend code which is bundled into thirdparty code is to be seen as GPLified.

For those facing issues with GPL projects, I usually do not distribute the GPL code but expect it to be there. And then (via installer or scripts) pull the non-gpl code. This is ofc not working for frontendcode which itselfs bundles gpl+nongpl code in one file as it is distributed together. For directus I did not have the case, yet.

**Reply by rijkvanzanten** [reply]
> For those facing issues with GPL projects, I usually do not distribute the GPL code but expect it to be there. And then (via installer or scripts) pull the non-gpl code. This is ofc not working for frontendcode which itselfs bundles gpl+nongpl code in one file as it is distributed together.

This is how I understand it as well. Using Directus should be just fine as you're not bundling/actively re-distributing the code? IANAL

**Reply by DracoBlue** [reply]
> > For those facing issues with GPL projects, I usually do not distribute the GPL code but expect it to be there. And then (via installer or scripts) pull the non-gpl code. This is ofc not working for frontendcode which itselfs bundles gpl+nongpl code in one file as it is distributed together.
> 
> This is how I understand it as well. Using Directus should be just fine as you're not bundling/actively re-distributing the code? IANAL

As directus/sdk is MIT https://github.com/directus/sdk/commit/6778d6694ac79fa1c1ff7e8061d967a5013f66eb - this is also true for everything build with it as long as you don’t distribute this as one package together with directus to your client in my understanding,  though IANAL.

For extensions bundled via „directus-extension build“ command, taken e.g. this require to the directus extension source code: https://github.com/dimitrov-adrian/directus-extension-editorjs-interface/blob/53057131ad6085e83aa35e28086b55029ee1da16/src/index.ts#L1  makes in my point of view of a non-lawyer the final build (an advertised way how to install or bundle extensions) to GPL. Just my understanding even though IANAL.

**Reply by DracoBlue** [reply]
I think with BUSL for directus and MIT for extensions this is much clearer now.

Thanks!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
